### PR TITLE
gh-91102: Port 8-argument _warnings.warn_explicit to Argument Clinic

### PR DESCRIPTION
**Pull request by arhadthedev**
Benefits:

1. `METH_VARARGS` calling convention is replaced with a little faster `METH_FASTCALL` (no extra tuple creation)
2. `PyArg_ParseTupleAndKeywords` call is replaced with `_PyArg_UnpackKeywords` and a tailored parser generated specially for the ported function.
3. Stresses that `warn_explicit()` differentiates situations when its `object` argument is set to `Py_None` and `NULL`, while for other optional arguments there is no difference.

/cc @rhettinger as a person who helped me to learn to classify whether extra abstractions pay off with benefits. This case looks like they do.

Related issue: gh-91102.

### CODE REVIEW

File: Python/_warnings.c
Comment (erlend-aasland) [inline Python/_warnings.c]:
```suggestion
    module: object = NULL
```

**Review by erlend-aasland — CHANGES_REQUESTED** [review]

File: Python/_warnings.c
Comment (arhadthedev) [reply Python/_warnings.c]: @erlend-aasland The `module` parameter is reserved by Argument Clinic and points to a parent module object:

```cpp
static PyObject *
warnings_warn_explicit_impl(PyObject *module, PyObject *message,
                            PyObject *category, PyObject *filename,
                            int lineno, PyObject *mod, PyObject *registry,
                            PyObject *module_globals, PyObject *sourceobj)
```

Comment (erlend-aasland) [reply Python/_warnings.c]: Then we should fix that first. This PR is currently a change in behaviour; one keyword argument is now named differently than before.

Comment (arhadthedev) [reply Python/_warnings.c]: I believe that a risk to break anything is next to zero:

- `warnings.warn_explicit` [performs lots of useful checks](https://github.com/python/cpython/blob/main/Lib/warnings.py#L327-L395) before calling `_warnings.warn_explicit` so users have no benefit to plug into the private implementation directly,
- the _warnings module is a private, undocumented backend.

> Then we should fix that first

The problem is not creation of the issue. The problem is that it'll stuck because the module arg renaming feature brings no use outside of a single PR.

Comment (kumaraditya303) [inline Python/_warnings.c]:
```suggestion
    module as mod: object = NULL
```
This should work or am I missing something ?

Comment (arhadthedev) [reply Python/_warnings.c]: > Error in file "Python/_warnings.c" on line 1053:
> You can't have two parameters named 'module'!

Comment (kumaraditya303) [reply Python/_warnings.c]: That sounds like a clinic bug, it should compare c level names as `module` is implicitly positional only as the first arg.

Comment (arhadthedev) [reply Python/_warnings.c]: I've sorted out that `module` is converted into `self` on early stages, so I've commited whatever is necessary.

The only caveat is two same named `module` parameters in a docstring, but I believe it's quite of no importance.

**Comment by arhadthedev** [comment]
I have made the requested changes; please review again.

File: Python/_warnings.c
Comment (erlend-aasland) [reply Python/_warnings.c]: As I said in https://github.com/python/cpython/pull/92891#discussion_r875466153, I still believe we should fix the AC bug first, and then proceed with this PR. Instead of adding hacks to work around limitations in AC, we should fix AC.

I suggest opening an issue on the AC limitation/bug (unless there is one already) and perhaps even proposing a PR to fix it.

Comment (arhadthedev) [reply Python/_warnings.c]: - gh-94246

Comment (erlend-aasland) [reply Python/_warnings.c]: @arhadthedev: the issue you opened does not address the problems with this PR.

**Comment by arhadthedev** [comment]
Turning this PR draft until gh-94431 is merged. Without that fix, a proper renaming (`module as _module`) doesn't work.

**Comment by erlend-aasland** [comment]
> Turning this PR draft until [gh-94431](https://github.com/python/cpython/pull/94431) is merged. Without that fix, a proper renaming (`module as _module`) doesn't work.

...given my proposed solution is an acceptable solution ;) Other core devs might not agree with me, so we might have to go with your workaround. IMO, gh-94431 is a good approach.

**Comment by arhadthedev** [comment]
@erlend-aasland Thank you for the fix applied; I renamed the second `module` as @kumaraditya303 suggested.

**Review by erlend-aasland — APPROVED** [review]
LGTM, but the NEWS entry is IMO too verbose. It should be sufficient to say that `_warnings.warn_explicit` is ported to AC, and nothing else. I don't think the nitty-gritty details are very interesting for the average user; for core devs, triagers, and other contributors, the benefits are obvious.

**Review by kumaraditya303 — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thanks, Oleg and Kumar.

**Pull request merged**
